channeldb/payment_status.go

```diff
@@ -161,7 +184,131 @@ func (ps PaymentStatus) Registrable(settled, paymentFailed bool) error {
+//	|   false  |   true  |  false |   no   |    StatusSucceeded   |
+//	|   false  |   false |  true  |   yes  |      StatusFailed    |
+//	|   false  |   false |  true  |   no   |    StatusInFlight    |
+//	|   false  |   false |  false |   yes  |      StatusFailed    |
+//	|   false  |   false |  false |   no   |    StatusInitiated   |
+//
+// Overall, `settled` and `reason` cannot be both true otherwise initiating an
+// unexpected condition. Also, when `inflight`, `settled`, and `failed` are
+// false, this initiates the payment is newly created and hasn't made any HTLCs
+// yet.
+func decidePaymentStatus(htlcs []HTLCAttempt,
+	reason *FailureReason) PaymentStatus {
+
+	var (
+		inflight      bool
+		htlcSettled   bool
+		htlcFailed    bool
+		paymentFailed bool
+	)
+
+	// If we have a failure reason, the payment is failed.
+	if reason != nil {
+		paymentFailed = true
+	}
+
+	// Go through all HTLCs for this payment, check whether we have any
+	// settled HTLC, and any still in-flight.
+	for _, h := range htlcs {
+		if h.Failure != nil {
+			htlcFailed = true
+			continue
+		}
+
+		if h.Settle != nil {
+			htlcSettled = true
+			continue
+		}
+
+		// If any of the HTLCs are not failed nor settled, we
+		// still have inflight HTLCs.
+		inflight = true
+	}
+
+	// Use the DB state to determine the status of the payment.
+	switch {
+	// If any of the HTLCs is settled yet the payment is marked as failed,
+	// this initiates an error state.
+	//
+	// NOTE: whether we have failed or inflight HTLCs is irrelevant.
+	case htlcSettled && paymentFailed:
+		return StatusError
+
+	// If we have inflight HTLCs, no matter we have settled or failed
+	// HTLCs, or the payment failed, we still consider it inflight so we
+	// inform upper systems to wait for the results.
+	case inflight:
+		return StatusInFlight
+
+	// If we have no in-flight HTLCs, and at least one of the HTLCs is
+	// settled, the payment succeeded.
+	//
+	// NOTE: when reaching this case, paymentFailed must be false.
+	case htlcSettled:
+		return StatusSucceeded
+
+	// If we have no in-flight HTLCs, and the payment failure is set, the
+	// payment is considered failed.
+	//
+	// NOTE: when reaching this case, settled must be false.
+	case paymentFailed:
+		return StatusFailed
+
+	// If we have no in-flight HTLCs, yet the payment is NOT failed, it
+	// means all the HTLCs are failed. In this case we can attempt more
+	// HTLCs.
+	//
+	// NOTE: when reaching this case, both settled and paymentFailed must
+	// be false.
+	case htlcFailed:
+		return StatusInFlight
+
+	// If none of the HTLCs is either settled or failed, and we have no
+	// inflight HTLCs, this means the payment has no HTLCs created yet.
+	//
+	// NOTE: when reaching this case, both settled and paymentFailed must
+	// be false.
+	case !htlcFailed:
+		return StatusInitiated
+
+	// Otherwise an impossible state is reached.
+	//
+	// NOTE: we should never end here.
+	default:
+		return StatusError
```

Comment:
Maybe panic is appropriate, because the last two conditions are `htlcFailed` and `!htlcFailed`. It's even more obvious that it can't ever happen then a simple array out of bounds for example which also panics.

Reply:
added an error log instead and think we'd avoid panicking, maybe remove this `default` case?

Reply:
I still think that when it is trivial to see that something should never happen, a panic is appropriate. But some personal preference in that I think.

Reply:
yeah it should never happen. my concern is, if it happens, the node would not be able to recover unless some manual db cleanup is performed...